src/dftbp/math/sparseblas.F90

```diff
@@ -0,0 +1,1183 @@
+    module procedure symm_bc_${suffix}$_gamma
+  #:endfor
+    module procedure symm_kpt
+    module procedure symm_bc_kpt
+  end interface symm
+
+contains
+
+#:for typename, suffix in ROUTINE_KINDS
+
+  !> Sparse Gamma point matrix with a ${typename}$ vector as a symv operation,
+  !> y = alpha A x + beta * y
+  subroutine symv_${suffix}$_gamma(y, A, x, iNeighbour, nNeighbour, img2CentCell, iSparseStart,&
+      & iAtomStart, orb, alpha, beta)
+
+    !> Vector on return
+    ${typename}$(dp), intent(inout):: y(:)
+
+    !> Sparse matrix
+    real(dp), intent(in) :: A(:)
+
+    !> Vector on entry
+    ${typename}$(dp), intent(in) :: x(:)
+
+    !> Atom neighbour list
+    integer, intent(in) :: iNeighbour(0:,:)
+
+    !> Number of neighbours for atoms
+    integer, intent(in) :: nNeighbour(:)
+
+    !> Image to central cell mapping
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Sparse indexing
+    integer, intent(in) :: iSparseStart(0:,:)
+
+    !> Dense indexing
+    integer, intent(in) :: iAtomStart(:)
+
+    !> data type for atomic orbital information
+    type(TOrbitals), intent(in) :: orb
+
+    !> Scaling factor for A * x
+    ${typename}$(dp), optional, intent(in) :: alpha
+
+    !> Scaling factor for incoming y
+    ${typename}$(dp), optional, intent(in) :: beta
+
+    integer :: iOrig, ix, iy, jx, jy, iNeigh, nAtom, iAtom1, iAtom2, iAtom2f, nOrb1, nOrb2, ii
+    real(dp) :: sqrTmp(orb%mOrb, orb%mOrb)
+    ${typename}$(dp) :: alphaTmp
+
+    @:ASSERT(size(x) == size(y))
+
+    nAtom = size(iAtomStart)-1
+
+    if (present(alpha)) then
+      alphaTmp = alpha
+    else
+      alphaTmp = 1.0_dp
+    end if
+
+    if (present(beta)) then
+      y(:) = beta * y
+    else
+      y(:) = 0.0_dp
+    end if
+
+    !$OMP PARALLEL DO DEFAULT(SHARED) &
+    !$OMP &PRIVATE(iAtom1,nOrb1,ix,jx,iNeigh,sqrTmp,iAtom2,iAtom2f,nOrb2,iOrig,ii,iy,jy) &
+    !$OMP & REDUCTION(+:y) SCHEDULE(RUNTIME)
+    do iAtom1 = 1, nAtom
+      ix = iAtomStart(iAtom1)
+      jx = iAtomStart(iAtom1 + 1) -1
+      nOrb1 = jx - ix + 1
+      do iNeigh = 0, nNeighbour(iAtom1)
+        iOrig = iSparseStart(iNeigh,iAtom1) + 1
+        sqrTmp(:,:) = 0.0_dp
+        iAtom2 = iNeighbour(iNeigh, iAtom1)
+        iAtom2f = img2CentCell(iAtom2)
+        iy = iAtomStart(iAtom2f)
+        jy = iAtomStart(iAtom2f + 1) - 1
+        nOrb2 = jy - iy + 1
+        sqrTmp(:nOrb2,:nOrb1) = reshape(A(iOrig:iOrig+nOrb2*nOrb1-1), [nOrb2,nOrb1])
+        ! symmetrize on-diagonal blocks just in case
+        if (iAtom1 == iAtom2f) then
+          do ii = 1, nOrb1
+            sqrTmp(ii,ii+1:nOrb2) = sqrTmp(ii+1:nOrb2,ii)
+          end do
+        end if
+        y(iy:jy) = y(iy:jy) + alphaTmp * matmul(sqrTmp(:nOrb2,:nOrb1),x(ix:jx))
+        ! other triangle due to symmetry of matrix
+        if (iAtom1 /= iAtom2f) then
+          y(ix:jx) = y(ix:jx) + alphaTmp * matmul(transpose(sqrTmp(:nOrb2,:nOrb1)),x(iy:jy))
+        end if
+      end do
+    end do
+    !$OMP  END PARALLEL DO
```

Comment:
This code block seems to be duplicated several times. Can you somehow reuse it?

Reply:
The _bc_ cases can also do the job of the other routines, but there will be a performance penalty. This will need benchmarking to decide once they get used for more operations in the code in a future PR.
Possibly the k-point case can also be used for non-k-point as well (but will need some refactoring in the main code to make sure all of the extra indexing arrays are consistently either allocated or not).